src/framework/linalg/linops/linops_aer_vector.hpp

```diff
@@ -0,0 +1,142 @@
+/**
+ * This code is part of Qiskit.
+ *
+ * (C) Copyright IBM 2018, 2019.
+ *
+ * This code is licensed under the Apache License, Version 2.0. You may
+ * obtain a copy of this license in the LICENSE.txt file in the root directory
+ * of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Any modifications or derivative works of this code must retain this
+ * copyright notice, and modified files need to carry a notice indicating
+ * that they have been altered from the originals.
+ */
+
+#ifndef _aer_framework_linalg_linops_aer_vector_hpp_
+#define _aer_framework_linalg_linops_aer_vector_hpp_
+
+#include <functional>
+#include <vector>
+
+#include "framework/linalg/almost_equal.hpp"
+#include "framework/linalg/vector.hpp"
+
+namespace AER {
+namespace Linalg {
+
+// This defines functions add, sub, mul, div and iadd, imul, isub, idiv
+// that work for numeric vector types
+
+//----------------------------------------------------------------------------
+// Linear operations
+//----------------------------------------------------------------------------
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T> add(const Vector<T>& lhs, const Vector<T>& rhs) {
+  return lhs + rhs;
+}
+
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T>& iadd(Vector<T>& lhs, const Vector<T>& rhs) {
+  lhs += rhs;
+  return lhs;
+}
+
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T> sub(const Vector<T>& lhs, const Vector<T>& rhs) {
+  return lhs - rhs;
+}
+
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T>& isub(Vector<T>& lhs, const Vector<T>& rhs) {
+  lhs -= rhs;
+  return lhs;
+}
+
+//----------------------------------------------------------------------------
+// Affine operations
+//----------------------------------------------------------------------------
+template <class T, class Scalar, typename = enable_if_numeric_t<T>,
```

Comment:
For these affine operations, in order to reduce code duplication, it would be better to implement the _op_ function in terms of the _iop_ equivalent one:
```
T op(const T& a, const Scalar& val){
   T tmp(a);
   return iop(a, val);
}
```